File: src/services/get_def/getDef_js.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Loc_collections

module Get_def_result = struct
  type t =
    | Def of LocSet.t * (* the final location of the definition, name *) string option
    | Partial of LocSet.t * (* name *) string option * string
        (** if an intermediate get-def failed, return partial progress and the error message *)
    | Bad_loc of string  (** the input loc didn't point at anything you can call get-def on *)
    | Def_error of string  (** an unexpected, internal error *)
end

open Get_def_result

let extract_member_def ~loc_of_aloc ~cx ~file_sig ~typed_ast_opt ~force_instance t name =
  let ( let* ) = Result.bind in
  let* Ty_members.{ members; _ } =
    Ty_members.extract
      ~allowed_prop_names:[Reason.OrdinaryName name]
      ~force_instance
      ~cx
      ~typed_ast_opt
      ~file_sig
      t
  in
  let def_locs =
    Base.Option.bind
      (NameUtils.Map.find_opt (Reason.OrdinaryName name) members)
      ~f:(fun Ty_members.{ def_locs; _ } -> Nel.of_list def_locs
    )
  in
  match def_locs with
  | Some def_locs -> Ok (Nel.map loc_of_aloc def_locs, Some name)
  | None -> Error (Printf.sprintf "failed to find member %s in members map" name)

let rec process_request ~loc_of_aloc ~cx ~is_legit_require ~ast ~typed_ast_opt ~file_sig ~scope_info
    : (ALoc.t, ALoc.t * Type.t) Get_def_request.t -> (Loc.t Nel.t * string option, string) result =
  function
  | Get_def_request.Identifier { name; loc = aloc } ->
    let loc = loc_of_aloc aloc in
    let all_uses = Scope_api.With_Loc.all_uses scope_info in
    let matching_uses = Loc_collections.LocSet.filter (fun use -> Loc.contains use loc) all_uses in
    (match Loc_collections.LocSet.elements matching_uses with
    | [use] ->
      let def = Scope_api.With_Loc.def_of_use scope_info use in
      let def_loc = def.Scope_api.With_Loc.Def.locs in
      Ok (def_loc, Some name)
    | [] ->
      (match Context.builtin_value_opt cx name with
      | Some (def_loc, _) -> Ok (Nel.one (loc_of_aloc def_loc), Some name)
      | None ->
        (match Context.builtin_type_opt cx name with
        | Some (def_loc, _) -> Ok (Nel.one (loc_of_aloc def_loc), Some name)
        | None -> Ok (Nel.one loc, Some name)))
    | _ :: _ :: _ -> Error "Scope builder found multiple matching identifiers")
  | Get_def_request.(Member { prop_name = name; object_type = (_loc, t); force_instance }) ->
    extract_member_def ~loc_of_aloc ~cx ~file_sig ~typed_ast_opt ~force_instance t name
  | Get_def_request.JsxAttribute { component_t = (_, component_t); name; loc } ->
    let reason = Reason.mk_reason Reason.(RProperty (Some (OrdinaryName name))) loc in
    let props_object =
      Tvar.mk_where cx reason (fun tvar ->
          let use_op = Type.Op Type.UnknownUse in
          Flow_js.flow
            cx
            (component_t, Type.ReactKitT (use_op, reason, Type.React.GetConfig { tout = tvar }))
      )
    in
    let req =
      Get_def_request.(
        Member { prop_name = name; object_type = (loc, props_object); force_instance = false }
      )
    in
    process_request ~loc_of_aloc ~cx ~is_legit_require ~ast ~typed_ast_opt ~file_sig ~scope_info req

module Depth = struct
  let limit = 100

  type error =
    | Cycle of Loc.t list
    | DepthExceeded of Loc.t list

  type t = {
    mutable length: int;
    mutable seen: Loc_collections.LocSet.t;
    mutable results: Get_def_result.t Loc_collections.LocMap.t;
    mutable locs: Loc.t list;
  }

  type ok =
    | NoResult
    | CachedResult of Get_def_result.t

  let empty () =
    {
      length = 0;
      seen = Loc_collections.LocSet.empty;
      locs = [];
      results = Loc_collections.LocMap.empty;
    }

  let add loc ({ length; seen; locs; results } as depth) =
    match Loc_collections.LocMap.find_opt loc results with
    | None ->
      let locs = loc :: locs in
      let seen' = Loc_collections.LocSet.add loc seen in
      if seen == seen' then
        Error (Cycle locs)
      else if length >= limit then
        Error (DepthExceeded locs)
      else (
        depth.seen <- seen';
        depth.length <- length + 1;
        depth.locs <- locs;
        Ok NoResult
      )
    | Some result -> Ok (CachedResult result)

  let cache_result loc result ({ results; _ } as depth) =
    depth.results <- Loc_collections.LocMap.add loc result results
end

exception FoundTokenAtRequestLoc of Token.t

let get_def ~loc_of_aloc ~cx ~file_sig ?file_content ~ast ~available_ast ~purpose requested_loc =
  let require_loc_map = File_sig.require_loc_map file_sig in
  let scope_info =
    Scope_builder.program ~enable_enums:(Context.enable_enums cx) ~with_types:true ast
  in
  let is_local_use aloc = Scope_api.With_Loc.is_local_use scope_info (loc_of_aloc aloc) in
  let is_legit_require source_aloc =
    let source_loc = loc_of_aloc source_aloc in
    Flow_import_specifier.Map.exists
      (fun _ locs -> List.exists (fun loc -> loc = source_loc) locs)
      require_loc_map
  in
  let rec loop ~scope_info ~depth req_loc loop_name =
    match Depth.add req_loc depth with
    | Error error ->
      let trace_of_locs locs =
        locs |> Base.List.map ~f:Reason.string_of_loc |> Base.String.concat ~sep:"\n"
      in
      let log_message =
        match error with
        | Depth.DepthExceeded locs ->
          Printf.sprintf
            "GetDef_js loop depth exceeded %d. Trace (most recent first):\n%s"
            Depth.limit
            (trace_of_locs locs)
        | Depth.Cycle locs ->
          Printf.sprintf
            "GetDef_js cycle detected. Trace (most recent first):\n%s"
            (trace_of_locs locs)
      in
      Partial (LocSet.singleton req_loc, loop_name, log_message)
    | Ok (Depth.CachedResult result) -> result
    | Ok Depth.NoResult ->
      let open Get_def_process_location in
      let result =
        match
          process_location cx ~is_local_use ~is_legit_require ~available_ast ~purpose req_loc
        with
        | OwnNamedDef (aloc, name) -> Def (LocSet.singleton (loc_of_aloc aloc), Some name)
        | OwnUnnamedDef aloc -> Def (LocSet.singleton (loc_of_aloc aloc), None)
        | ModuleDef aloc -> Def (LocSet.singleton (loc_of_aloc aloc), None)
        | ModuleTypeDef t ->
          (match Get_def_process_location.process_type_request cx t with
          | Ok res_loc -> Def (LocSet.singleton (loc_of_aloc res_loc), None)
          | Error e -> Def_error e)
        | Request request -> begin
          match
            process_request
              ~loc_of_aloc
              ~cx
              ~is_legit_require
              ~ast
              ~typed_ast_opt:(Typed_ast_utils.typed_ast_of_available_ast available_ast)
              ~file_sig
              ~scope_info
              request
          with
          | Ok (res_locs, name) ->
            res_locs
            |> Nel.map (fun res_loc ->
                   (* two scenarios where we stop trying to recur:
                      - when req_loc = res_loc, meaning we've reached a fixed point so
                        continuing would make us infinite loop
                      - when res_loc is not in the file the request originated from, meaning
                        the typed_ast we have is the wrong one to recur with *)
                   if Loc.equal req_loc res_loc || Loc.(res_loc.source <> requested_loc.source) then
                     Def (LocSet.singleton res_loc, name)
                   else
                     match loop ~scope_info ~depth res_loc name with
                     | Bad_loc _ -> Def (LocSet.singleton res_loc, name)
                     | Def_error msg -> Partial (LocSet.singleton res_loc, name, msg)
                     | (Def _ | Partial _) as res -> res
               )
            |> Nel.reduce (fun res1 res2 ->
                   match (res1, res2) with
                   | (Def (locs1, n), Def (locs2, _)) -> Def (LocSet.union locs1 locs2, n)
                   | (Partial (locs1, n, msg1), Partial (locs2, _, msg2)) ->
                     Partial (LocSet.union locs1 locs2, n, msg1 ^ msg2)
                   | (Def (locs1, n), Partial (locs2, _, msg))
                   | (Partial (locs1, n, msg), Def (locs2, _)) ->
                     Partial (LocSet.union locs1 locs2, n, msg)
                   | ((Bad_loc _ | Def_error _), other)
                   | (other, (Bad_loc _ | Def_error _)) ->
                     other
               )
          | Error msg -> Def_error msg
        end
        | Empty msg -> Bad_loc msg
        | LocNotFound ->
          let token_at_req_loc =
            match file_content with
            | None -> None
            | Some file_content ->
              let token_sink =
                Some
                  (fun { Parser_env.token_loc; token; _ } ->
                    if Loc.contains token_loc req_loc then raise (FoundTokenAtRequestLoc token))
              in
              (try
                 ignore
                 @@ Parser_flow.program_file
                      ~token_sink
                      ~fail:false
                      ~parse_options:(Some Parser_env.permissive_parse_options)
                      file_content
                      (Some (Context.file cx));
                 None
               with
              | FoundTokenAtRequestLoc token -> Some token)
          in
          (match token_at_req_loc with
          | Some token ->
            Bad_loc
              (Utils_js.spf
                 "unsupported token: %s (%s)"
                 (Token.token_to_string token)
                 (Token.value_of_token token)
              )
          | None -> Bad_loc "not found")
        | InternalError err -> Def_error (Get_def_process_location.show_internal_error err)
      in
      Depth.cache_result req_loc result depth;
      result
  in
  loop ~depth:(Depth.empty ()) ~scope_info requested_loc None

File: rust_port/crates/flow_services_get_def/src/get_def_js.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::BTreeSet;

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_analysis::scope_api;
use flow_analysis::scope_builder;
use flow_common::reason;
use flow_common::reason::Name;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::PERMISSIVE_PARSE_OPTIONS;
use flow_parser::ast;
use flow_parser::loc::Loc;
use flow_parser::token::TokenKind;
use flow_parser_utils::file_sig::FileSig;
use flow_typing_context::Context;
use flow_typing_type::type_::Type;
use flow_typing_utils::typed_ast_utils::AvailableAst;
use vec1::Vec1;

use crate::get_def_process_location;
use crate::get_def_request::GetDefRequest;
use crate::get_def_request::MemberInfo;
use crate::get_def_types::Purpose;

#[derive(Debug, Clone)]
pub enum GetDefResult {
    Def(BTreeSet<Loc>, Option<FlowSmolStr>), // (the final location of the definition, name)
    /// if an intermediate get-def failed, return partial progress and the error message
    Partial(BTreeSet<Loc>, /* name */ Option<FlowSmolStr>, String),
    /// the input loc didn't point at anything you can call get-def on
    BadLoc(String),
    /// an unexpected, internal error
    DefError(String),
}

fn extract_member_def<'cx>(
    loc_of_aloc: &dyn Fn(&ALoc) -> Loc,
    cx: &Context<'cx>,
    file_sig: &std::sync::Arc<FileSig>,
    typed_ast_opt: Option<&ast::Program<ALoc, (ALoc, Type)>>,
    force_instance: bool,
    t: &Type,
    name: &FlowSmolStr,
) -> Result<(Vec1<Loc>, Option<FlowSmolStr>), String> {
    let extract_result = flow_typing::ty_members::extract(
        force_instance,
        Some(vec![Name::new(name.dupe())]),
        None,
        cx,
        typed_ast_opt,
        file_sig.dupe(),
        t,
    )?;

    let key = Name::new(name.dupe());
    let def_locs = extract_result
        .members
        .get(&key)
        .and_then(|info| Vec1::try_from_vec(info.def_locs.clone()).ok());

    match def_locs {
        Some(def_locs) => {
            let mapped = def_locs.mapped_ref(loc_of_aloc);
            Ok((mapped, Some(name.dupe())))
        }
        None => Err(format!("failed to find member {} in members map", name)),
    }
}

fn process_request<'cx>(
    loc_of_aloc: &dyn Fn(&ALoc) -> Loc,
    cx: &Context<'cx>,
    typed_ast_opt: Option<&ast::Program<ALoc, (ALoc, Type)>>,
    file_sig: &std::sync::Arc<FileSig>,
    scope_info: &scope_api::ScopeInfo<Loc>,
    request: &GetDefRequest<ALoc, (ALoc, Type)>,
) -> Result<
    Result<(Vec1<Loc>, Option<FlowSmolStr>), String>,
    flow_utils_concurrency::job_error::JobError,
> {
    match request {
        GetDefRequest::Identifier { name, loc: aloc } => {
            let loc = loc_of_aloc(aloc);
            let all_uses = scope_info.all_uses();
            let matching_uses: Vec<&Loc> =
                all_uses.into_iter().filter(|u| u.contains(&loc)).collect();
            match matching_uses.len() {
                1 => {
                    let use_loc = matching_uses[0];
                    let def = scope_info
                        .def_of_use_opt(use_loc)
                        .expect("use should have a def");
                    let def_locs = def.locs.clone();
                    Ok(Ok((def_locs, Some(name.dupe()))))
                }
                0 => match cx.builtin_value_opt(name) {
                    Some((def_loc, _)) => {
                        Ok(Ok((Vec1::new(loc_of_aloc(&def_loc)), Some(name.dupe()))))
                    }
                    None => match cx.builtin_type_opt(name) {
                        Some((def_loc, _)) => {
                            Ok(Ok((Vec1::new(loc_of_aloc(&def_loc)), Some(name.dupe()))))
                        }
                        None => Ok(Ok((Vec1::new(loc), Some(name.dupe())))),
                    },
                },
                _ => Ok(Err(
                    "Scope builder found multiple matching identifiers".to_string()
                )),
            }
        }

        GetDefRequest::Member(member_info) => {
            let name = &member_info.prop_name;
            let (_, t) = &member_info.object_type;
            let force_instance = member_info.force_instance;
            Ok(extract_member_def(
                loc_of_aloc,
                cx,
                file_sig,
                typed_ast_opt,
                force_instance,
                t,
                name,
            ))
        }

        GetDefRequest::JsxAttribute {
            component_t: (_, component_t),
            name,
            loc,
        } => {
            use flow_common::reason::VirtualReasonDesc::*;
            use flow_typing_type::type_::*;

            let reason = reason::mk_reason(RProperty(Some(Name::new(name.dupe()))), loc.dupe());
            let props_object = flow_typing_tvar::mk_where(cx, reason.dupe(), |cx, tvar| {
                let use_op = UseOp::Op(std::sync::Arc::new(VirtualRootUseOp::UnknownUse));
                let use_t = UseT::new(UseTInner::ReactKitT(Box::new(ReactKitTData {
                    use_op,
                    reason: reason.dupe(),
                    tool: Box::new(react::Tool::GetConfig { tout: tvar.dupe() }),
                })));
                flow_typing_flow_js::flow_js::flow_non_speculating(cx, (component_t, &use_t))
            })?;
            let req = GetDefRequest::Member(MemberInfo {
                prop_name: name.dupe(),
                object_type: (loc.dupe(), props_object),
                force_instance: false,
            });
            process_request(loc_of_aloc, cx, typed_ast_opt, file_sig, scope_info, &req)
        }
    }
}

mod depth {
    use super::*;

    pub const LIMIT: usize = 100;

    pub enum DepthError {
        Cycle(Vec<Loc>),
        DepthExceeded(Vec<Loc>),
    }

    pub struct Depth {
        pub length: usize,
        pub seen: BTreeSet<Loc>,
        pub results: BTreeMap<Loc, GetDefResult>,
        pub locs: Vec<Loc>,
    }

    pub enum DepthOk {
        NoResult,
        CachedResult(GetDefResult),
    }

    impl Depth {
        pub fn empty() -> Self {
            Depth {
                length: 0,
                seen: BTreeSet::new(),
                results: BTreeMap::new(),
                locs: Vec::new(),
            }
        }

        pub fn add(&mut self, loc: &Loc) -> Result<DepthOk, DepthError> {
            if let Some(result) = self.results.get(loc) {
                return Ok(DepthOk::CachedResult(result.clone()));
            }
            let mut new_locs = vec![loc.clone()];
            new_locs.extend(self.locs.iter().cloned());
            if self.seen.contains(loc) {
                return Err(DepthError::Cycle(new_locs));
            }
            if self.length >= LIMIT {
                return Err(DepthError::DepthExceeded(new_locs));
            }
            self.seen.insert(loc.clone());
            self.length += 1;
            self.locs = new_locs;
            Ok(DepthOk::NoResult)
        }

        pub fn cache_result(&mut self, loc: Loc, result: GetDefResult) {
            self.results.insert(loc, result);
        }
    }
}

// exception FoundTokenAtRequestLoc of Token.t

pub fn get_def<'cx>(
    loc_of_aloc: &dyn Fn(&ALoc) -> Loc,
    cx: &Context<'cx>,
    file_sig: &std::sync::Arc<FileSig>,
    file_content: Option<&str>,
    ast: &ast::Program<Loc, Loc>,
    available_ast: AvailableAst,
    purpose: &Purpose,
    requested_loc: &Loc,
) -> Result<GetDefResult, flow_utils_concurrency::job_error::JobError> {
    // Note: The Result here propagates Worker_should_cancel/timeout via JobError.
    // Both inner errors (from process_request) propagate via the Result chain.
    let require_loc_map = file_sig.require_loc_map();
    let scope_info = scope_builder::program(cx.enable_enums(), true, ast);
    let is_local_use = |aloc: &ALoc| -> bool {
        let loc = loc_of_aloc(aloc);
        scope_info.is_local_use(&loc)
    };
    let is_legit_require = |source_aloc: &ALoc| -> bool {
        let source_loc = loc_of_aloc(source_aloc);
        require_loc_map
            .values()
            .any(|locs| locs.contains(&source_loc))
    };

    let typed_ast_opt = match &available_ast {
        AvailableAst::TypedAst(tast) => Some(tast),
        AvailableAst::ALocAst(_) => None,
    };

    fn loop_fn<'cx>(
        loc_of_aloc: &dyn Fn(&ALoc) -> Loc,
        cx: &Context<'cx>,
        is_legit_require: &dyn Fn(&ALoc) -> bool,
        typed_ast_opt: Option<&ast::Program<ALoc, (ALoc, Type)>>,
        file_sig: &std::sync::Arc<FileSig>,
        scope_info: &scope_api::ScopeInfo<Loc>,
        available_ast: &AvailableAst,
        file_content: Option<&str>,
        is_local_use: &dyn Fn(&ALoc) -> bool,
        purpose: &Purpose,
        requested_loc: &Loc,
        depth_state: &mut depth::Depth,
        req_loc: &Loc,
        loop_name: Option<FlowSmolStr>,
    ) -> Result<GetDefResult, flow_utils_concurrency::job_error::JobError> {
        match depth_state.add(req_loc) {
            Err(error) => {
                let trace_of_locs = |locs: &Vec<Loc>| -> String {
                    locs.iter()
                        .map(|l| reason::string_of_loc(None, l))
                        .collect::<Vec<_>>()
                        .join("\n")
                };
                let log_message = match &error {
                    depth::DepthError::DepthExceeded(locs) => format!(
                        "GetDef_js loop depth exceeded {}. Trace (most recent first):\n{}",
                        depth::LIMIT,
                        trace_of_locs(locs)
                    ),
                    depth::DepthError::Cycle(locs) => format!(
                        "GetDef_js cycle detected. Trace (most recent first):\n{}",
                        trace_of_locs(locs)
                    ),
                };
                let mut locs = BTreeSet::new();
                locs.insert(req_loc.dupe());
                Ok(GetDefResult::Partial(locs, loop_name, log_message))
            }
            Ok(depth::DepthOk::CachedResult(result)) => Ok(result),
            Ok(depth::DepthOk::NoResult) => {
                let process_result = get_def_process_location::process_location(
                    cx,
                    available_ast,
                    is_local_use,
                    is_legit_require,
                    *purpose,
                    req_loc.clone(),
                );
                let result = match process_result {
                    get_def_process_location::ProcessLocationResult::OwnNamedDef(aloc, name) => {
                        let mut locs = BTreeSet::new();
                        locs.insert(loc_of_aloc(&aloc));
                        GetDefResult::Def(locs, Some(name))
                    }
                    get_def_process_location::ProcessLocationResult::OwnUnnamedDef(aloc) => {
                        let mut locs = BTreeSet::new();
                        locs.insert(loc_of_aloc(&aloc));
                        GetDefResult::Def(locs, None)
                    }
                    get_def_process_location::ProcessLocationResult::ModuleDef(aloc) => {
                        let mut locs = BTreeSet::new();
                        locs.insert(loc_of_aloc(&aloc));
                        GetDefResult::Def(locs, None)
                    }
                    get_def_process_location::ProcessLocationResult::ModuleTypeDef(t) => {
                        match get_def_process_location::process_type_request(cx, &t) {
                            Ok(res_loc) => {
                                let mut locs = BTreeSet::new();
                                locs.insert(loc_of_aloc(&res_loc));
                                GetDefResult::Def(locs, None)
                            }
                            Err(e) => GetDefResult::DefError(e),
                        }
                    }
                    get_def_process_location::ProcessLocationResult::Request(request) => {
                        let inner = process_request(
                            loc_of_aloc,
                            cx,
                            typed_ast_opt,
                            file_sig,
                            scope_info,
                            &request,
                        )?;
                        match inner {
                            Ok((res_locs, name)) => {
                                let mut results: Vec<GetDefResult> = Vec::new();
                                for res_loc in res_locs.iter() {
                                    let r = if *res_loc == *req_loc
                                        || res_loc.source != requested_loc.source
                                    {
                                        let mut locs = BTreeSet::new();
                                        locs.insert(res_loc.clone());
                                        GetDefResult::Def(locs, name.dupe())
                                    } else {
                                        match loop_fn(
                                            loc_of_aloc,
                                            cx,
                                            is_legit_require,
                                            typed_ast_opt,
                                            file_sig,
                                            scope_info,
                                            available_ast,
                                            file_content,
                                            is_local_use,
                                            purpose,
                                            requested_loc,
                                            depth_state,
                                            res_loc,
                                            name.dupe(),
                                        )? {
                                            GetDefResult::BadLoc(_) => {
                                                let mut locs = BTreeSet::new();
                                                locs.insert(res_loc.clone());
                                                GetDefResult::Def(locs, name.dupe())
                                            }
                                            GetDefResult::DefError(msg) => {
                                                let mut locs = BTreeSet::new();
                                                locs.insert(res_loc.clone());
                                                GetDefResult::Partial(locs, name.dupe(), msg)
                                            }
                                            res @ (GetDefResult::Def(..)
                                            | GetDefResult::Partial(..)) => res,
                                        }
                                    };
                                    results.push(r);
                                }
                                results
                                    .into_iter()
                                    .reduce(|res1, res2| match (res1, res2) {
                                        (
                                            GetDefResult::Def(locs1, n),
                                            GetDefResult::Def(locs2, _),
                                        ) => {
                                            let locs = &locs1 | &locs2;
                                            GetDefResult::Def(locs, n)
                                        }
                                        (
                                            GetDefResult::Partial(locs1, n, msg1),
                                            GetDefResult::Partial(locs2, _, msg2),
                                        ) => {
                                            let locs = &locs1 | &locs2;
                                            GetDefResult::Partial(
                                                locs,
                                                n,
                                                format!("{}{}", msg1, msg2),
                                            )
                                        }
                                        (
                                            GetDefResult::Def(locs1, n),
                                            GetDefResult::Partial(locs2, _, msg),
                                        )
                                        | (
                                            GetDefResult::Partial(locs1, n, msg),
                                            GetDefResult::Def(locs2, _),
                                        ) => {
                                            let locs = &locs1 | &locs2;
                                            GetDefResult::Partial(locs, n, msg)
                                        }
                                        (
                                            GetDefResult::BadLoc(_) | GetDefResult::DefError(_),
                                            other,
                                        )
                                        | (
                                            other,
                                            GetDefResult::BadLoc(_) | GetDefResult::DefError(_),
                                        ) => other,
                                    })
                                    .expect("results is non-empty because res_locs is Vec1")
                            }
                            Err(msg) => GetDefResult::DefError(msg),
                        }
                    }
                    get_def_process_location::ProcessLocationResult::Empty(msg) => {
                        GetDefResult::BadLoc(msg.into())
                    }
                    get_def_process_location::ProcessLocationResult::LocNotFound => {
                        let token_at_req_loc = match file_content {
                            None => None,
                            Some(file_content) => {
                                let mut found_token: Option<TokenKind> = None;
                                let mut token_sink = |result: flow_parser::TokenSinkResult| {
                                    if found_token.is_none() && result.token_loc.contains(req_loc) {
                                        found_token = Some(result.token_kind);
                                    }
                                };
                                flow_parser::parse_program_file::<()>(
                                    false,
                                    Some(&mut token_sink),
                                    Some(PERMISSIVE_PARSE_OPTIONS),
                                    cx.file().dupe(),
                                    Ok(file_content),
                                );
                                found_token
                            }
                        };
                        match token_at_req_loc {
                            Some(token) => {
                                GetDefResult::BadLoc(format!("unsupported token: {:?}", token))
                            }
                            None => GetDefResult::BadLoc("not found".to_string()),
                        }
                    }
                    get_def_process_location::ProcessLocationResult::InternalError(err) => {
                        GetDefResult::DefError(format!("{:?}", err))
                    }
                };
                depth_state.cache_result(req_loc.clone(), result.clone());
                Ok(result)
            }
        }
    }

    let mut depth_state = depth::Depth::empty();
    loop_fn(
        loc_of_aloc,
        cx,
        &is_legit_require,
        typed_ast_opt,
        file_sig,
        &scope_info,
        &available_ast,
        file_content,
        &is_local_use,
        purpose,
        requested_loc,
        &mut depth_state,
        requested_loc,
        None,
    )
}
